target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/etherdevice.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mtd/mtd.h>
+#include <linux/nvmem-consumer.h>
+#include <linux/nvmem-provider.h>
+#include <linux/platform_device.h>
+
+#define RB_MAGIC_HARD		"Hard"
+#define RB_HARDCONFIG_LEN 	0x1000
+
+struct hard_config_header {
+	u8 magic[4];
+} __packed;
+
+struct hard_config_tlv {
+	u16 tag;
+	u16 len;
+} __packed;
```

Comment:
I needed to swap these when big endian (ath79) (in my own very rough driver that does the same thing).
```
OFFSET      mikrotik/OEM/hapac2/mtd/hapac2_c454.img
--------------------------------------------------------------------------------
0x0008E000  48 61 72 64 1A 00 04 00 00 00 00 00 04 00 08 00 |Hard............|
0x0008E010  48 8F 5A
```
```
OFFSET      mikrotik/OEM/hapac/mtd/RouterBoot.img
--------------------------------------------------------------------------------
0x0000E000  64 72 61 48 00 04 00 1A 00 00 00 00 00 08 00 04 |draH............|
0x0000E010  48 8F 5A
```

Reply:
Hm, I based this implementation on the existing platform drivers and it claimed that the hard_config content is cpu-endian.

I still need to dig out some MIPS board to check the situation.

Reply:
hard-config tags are CPU-endian, yes: https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/routerboot.c#L37